### ISSUE TITLE
Multi-gpu example freeze and is not killable

### ISSUE THREAD

**Issue by Dubrzr** [pytorch/pytorch#24081]
## 🐛 Bug

Running pytorch with multiple P40 gpus freeze and is not killable (even kill -9 by root). Only a reboot removes this process.

Inside docker container (with nvidia-docker2) it freezes docker. https://github.com/NVIDIA/nvidia-docker/issues/1010 

## To Reproduce

1. Install pytorch 1.0.2
2. Run the following code on multiple P40 Gpus

```
import os


###tutorial from https://pytorch.org/tutorials/beginner/blitz/data_parallel_tutorial.html
###no error with only 1 gpu
# os.environ['CUDA_VISIBLE_DEVICES'] = '0'

#### to reproduce error allow multi gpu
os.environ['CUDA_VISIBLE_DEVICES'] = '0,1,2,3'

import torch


torch.cuda.device_count()

import torch.nn as nn
from torch.utils.data import Dataset, DataLoader

# Parameters and DataLoaders
input_size = 5000 #increased input size (works with 500 on multi gpu)
output_size = 2000 #increased output size (works with 200 on multi gpu)

batch_size = 300
data_size = 100

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print(device)

class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len

rand_loader = DataLoader(dataset=RandomDataset(input_size, data_size),
                         batch_size=batch_size, shuffle=True)

class Model(nn.Module):
    # Our model

    def __init__(self, input_size, output_size):
        super(Model, self).__init__()
        self.fc = nn.Linear(input_size, output_size)

    def forward(self, input):
        output = self.fc(input)


        return output

model = Model(input_size, output_size)
if torch.cuda.device_count() > 1:
  print("Let's use", torch.cuda.device_count(), "GPUs!")
  # dim = 0 [30, xxx] -> [10, ...], [10, ...], [10, ...] on 3 GPUs
  model = nn.DataParallel(model)

model.to(device)

for i in range(10000):
    for data in rand_loader:
        input = data.to(device)
        output = model(input)

```

## Expected behavior

The training

## Environment

Collecting environment information...
PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.2 LTS
GCC version: (crosstool-NG fa8859cb) 7.2.0
CMake version: Could not collect

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: Tesla P40
GPU 1: Tesla P40
GPU 2: Tesla P40
GPU 3: Tesla P40
GPU 4: Tesla P40
GPU 5: Tesla P40
GPU 6: Tesla P40
GPU 7: Tesla P40

Nvidia driver version: 410.79
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip3] numpy==1.15.2
[conda] mkl                       2018.0.3                      1    defaults
[conda] mkl_fft                   1.0.6                    py35_0    conda-forge
[conda] mkl_random                1.0.1                    py35_0    conda-forge
[conda] nomkl                     2.0                           0    defaults
[conda] numexpr                   2.6.5           py35_nomklhaa809a4_0  [nomkl]  defaults
[conda] pytorch                   1.0.1           py3.5_cuda10.0.130_cudnn7.4.2_2    pytorch
[conda] torch                     0.4.1                     <pip>
[conda] torchvision               0.2.2                      py_3    pytorch

**Comment by Dubrzr** [comment]
I've just tested with pytorch 1.2.0 it freezes in the same way.

**Comment by ezyang** [comment]
If someone wants to help us debug this issue, if you can try reproducing this issue that would be helpful; it would also be helpful to know if the problem goes away if you (1) upgrade your CUDA version and (2) run on different GPUs.

BTW, I notice in your environment, you have 

```
[conda] pytorch 1.0.1 py3.5_cuda10.0.130_cudnn7.4.2_2 pytorch
[conda] torch 0.4.1 
```

You should delete the old torch 0.4.1 at some point.

**Comment by rrkarim** [comment]
Same on `pytorch 1.1.0` with two `gtx 1080ti`.

**Comment by ezyang** [comment]
Upgrading priority

**Comment by raethlein** [comment]
We face a similar issue.

The environment below is with a single GPU mounted into the container, which is currently tested. Before, 2 GPUs were mounted in the container. When some GPU-specific code was run, such as `MyDecoder().to(device)`, the jupyter notebook server became completely unresponsive when it was tried to interrupt / restart the kernel. `docker stop` or `docker rm` did not work then anymore.
Only way to remove the container was to kill the container process, which left zombie and orphaned processes blocking the GPU which can only be fixed via a system restart.
It seems that the last log output is `... NotebookApp] Starting buffering for <kernel-id>`.

Though, I am not completely sure whether this issue is related or whether we have a hardware failure. Just thought it sounds so similar that I provide some information. I hope it helps.

**Environment**
Collecting environment information...
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.14.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0

Versions of relevant libraries:
[pip3] gpytorch==0.3.4
[pip3] msgpack-numpy==0.4.4.3
[pip3] numpy==1.16.4
[pip3] pytorch-ignite==0.2.0
[pip3] pytorch-nlp==0.4.1
[pip3] pytorch-pretrained-bert==0.6.2
[pip3] torch==1.1.0
[pip3] torch-cluster==1.4.3
[pip3] torch-geometric==1.3.0
[pip3] torch-scatter==1.3.1
[pip3] torch-sparse==0.4.0
[pip3] torchbiggraph==1.dev1
[pip3] torchfile==0.1.0
[pip3] torchstat==0.0.7
[pip3] torchtext==0.3.1
[pip3] torchvision==0.3.0
[conda] _tflow_select             2.3.0                       mkl    defaults
[conda] blas                      1.0                         mkl    defaults
[conda] faiss-cpu                 1.5.3            py36h6bb024c_0    pytorch
[conda] gpytorch                  0.3.4                    pypi_0    pypi
[conda] mkl                       2019.4                      243    defaults
[conda] mkl-include               2019.4                      243    defaults
[conda] mkl-service               2.0.2            py36h7b6447c_0    defaults
[conda] mkl_fft                   1.0.12           py36ha843d7b_0    defaults
[conda] mkl_random                1.0.2            py36hd81dba3_0    defaults
[conda] pytorch                   1.1.0           py3.6_cuda10.0.130_cudnn7.5.1_0    pytorch
[conda] pytorch-cpu               1.1.0               py3.6_cpu_0    pytorch
[conda] pytorch-ignite            0.2.0                    pypi_0    pypi
[conda] pytorch-nlp               0.4.1                    pypi_0    pypi
[conda] pytorch-pretrained-bert   0.6.2                    pypi_0    pypi
[conda] tensorflow                1.14.0          mkl_py36h2526735_0    defaults
[conda] torch-cluster             1.4.3                    pypi_0    pypi
[conda] torch-geometric           1.3.0                    pypi_0    pypi
[conda] torch-scatter             1.3.1                    pypi_0    pypi
[conda] torch-sparse              0.4.0                    pypi_0    pypi
[conda] torchbiggraph             1.dev1                   pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchstat                 0.0.7                    pypi_0    pypi
[conda] torchtext                 0.3.1                    pypi_0    pypi
[conda] torchvision               0.3.0           py36_cu10.0.130_1    pytorch
[conda] torchvision-cpu           0.3.0             py36_cuNone_1    pytorch

**Comment by Dubrzr** [comment]
I've realized a strace on this script : https://gist.github.com/Dubrzr/b058b54947b2688b7e02e64f6bdf78b8

It froze after those lines:

```
ioctl(4, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2a, 0x20), 0x7ffe185b07d0) = 0
futex(0x7fe8ba7a5e60, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x7fe8ba7c1ff8, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x55735fbc5450, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x55735fbc5454, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x55735fbc5458, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x55735fbc545c, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x7fe932502424, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x7fe92f71e520, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x7fe8f77a62e8, FUTEX_WAKE_PRIVATE, 2147483647) = 0
pipe2([12, 13], O_CLOEXEC)              = 0
fcntl(12, F_SETFL, O_RDONLY|O_NONBLOCK) = 0
clone(child_stack=0x7fe8cdef2fb0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tidptr=0x7fe8cdef39d0, tls=0x7fe8cdef3700, child_tidptr=0x7fe8cdef39d0) = 47266
ioctl(4, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2a, 0x20), 0x7ffe185af8d0) = 0
ioctl(5, _IOC(0, 0x00, 0x19, 0x00)
```

Hope this will help you :)

**Comment by rrkarim** [comment]
Is this one relatable: #1637? The solution: https://github.com/pytorch/pytorch/issues/1637#issuecomment-338268158

**Comment by jpellman** [comment]
I'm encountering the same issue as well- not even SIGKILL will work to stop a script using PyTorch.  Honestly, if SIGKILL isn't working, I doubt that this issue is specific to PyTorch.  I'd be more inclined to believe that the NVIDIA driver / tainted Linux kernel is at fault- anything outside the kernel space really shouldn't be able to mess up SIGKILL.  Similar issues have occurred in the past with _nvidia-smi_ (see [here](https://stackoverflow.com/questions/41489070/nvidia-smi-process-hangs-and-cant-be-killed-with-sigkill-either), [here](https://askubuntu.com/questions/753923/nvidia-smi-hangs-cannot-be-killed-even-by-sigkill), and [here](https://devtalk.nvidia.com/default/topic/929282/nvidia-smi-hangs-cannot-be-killed-even-by-sigkill/?offset=1)) 

For added context, my strace output is similar to @Dubrzr 's:

```
futex(0x7f6f6f90c1a8, FUTEX_WAKE_PRIVATE, 2147483647) = 0
pipe2([32, 33], O_CLOEXEC)              = 0
fcntl(32, F_SETFL, O_RDONLY|O_NONBLOCK) = 0
fcntl(33, F_SETFL, O_RDONLY|O_NONBLOCK) = 0
mmap(NULL, 8392704, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK, -1, 0) = 0x7f6ef5ef9000
mprotect(0x7f6ef5ef9000, 4096, PROT_NONE) = 0
clone(child_stack=0x7f6ef66f8fb0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tidptr=0x7f6ef66f99d0, tls=0x7f6ef66f9700, child_tidptr=0x7f6ef66f99d0) = 26182
futex(0x560c1db6f668, FUTEX_WAKE_PRIVATE, 1) = 1
ioctl(4, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2a, 0x20), 0x7ffe49a26910) = 0
ioctl(5, _IOC(0, 0x00, 0x19, 0x00)
```

The NVIDIA driver version is 418.87.00, cuda is 10.0.130, cudnn is 7.6.1.34-10.0, and PyTorch is 1.1.0.  I'm not the chief developer of the script producing the behavior so I don't know many details about how PyTorch is being used.

**Comment by jpellman** [comment]
For quick reference, here's a set of tallies for some of the relevant software/version number pairings:

## *NVIDIA Driver Summary Count:*

|Version|Count|
|-------|-------|
|410.79|1|
|418.67|1|
|418.87|1|

## *CUDA Summary Count:*

|Version|Count|
|-------|-------|
|10.0.130|3|

## *cudNN Summary Count:*

|Version|Count|
|-------|-------|
|7.4.2|1|
|7.6.0|1|
|7.6.1|1|

## *PyTorch Summary Count:*

|Version|Count|
|-------|-------|
|1.0.1.post2|1|
|1.2.0|1|
|1.1.0|3|

Based off these frequencies, CUDA is the most common factor.

**Comment by jpellman** [comment]
I was able to replicate this issue using an environment identical to my earlier one except for the CUDA version, which was 9.2.88.

**Comment by jpellman** [comment]
As another data point, the version of fastai exhibiting a similar issue that I linked to in my earlier comment would have also used CUDA 9.2 (based off a contemporary copy of the README [here|https://github.com/fastai/fastai/tree/ca55d51980719b524ba7a22d7cadbe53bdf6cfef]).

**Comment by z687wang** [comment]
Have exact same issue with 2 1080ti, CUDA 10 with pytorch stable (1.2)
pytorch installed via '''conda install pytorch torchvision cudatoolkit=10.0 -c pytorch'''

**Comment by jpellman** [comment]
A couple other tidbits about my configuration:

* IOMMU is not on, virtualization is disabled, both in software and hardware.
* I'm able to replicate the issue by just running _p2pBandwidthLatencyTest_ (one of the CUDA examples, alluded to at https://github.com/pytorch/pytorch/issues/1637#issuecomment-338268158; thanks @rrkarim ) alone so it's definitely not a PyTorch issue.
* The issue is intermittently reproducible with _p2pBandwidthLatencyTest_.  It seems to happen more regularly when two _p2pBandwidthLatencyTest_ processes are invoked at the same time.  When _p2pBandwidthLatencyTest_ runs as expected, I see the following in my _dmesg_ output:

```
[  970.469088] nvidia-uvm: Loaded the UVM driver in 8 mode, major device number 236
[  970.475874] nvidia 0000:1a:00.0: irq 279 for MSI/MSI-X
[  972.388240] nvidia 0000:1b:00.0: irq 280 for MSI/MSI-X
[  973.712312] nvidia 0000:3d:00.0: irq 281 for MSI/MSI-X
[  975.033118] nvidia 0000:3e:00.0: irq 282 for MSI/MSI-X
[  976.374481] nvidia 0000:88:00.0: irq 283 for MSI/MSI-X
[  977.731423] nvidia 0000:89:00.0: irq 284 for MSI/MSI-X
[  979.084570] nvidia 0000:b1:00.0: irq 285 for MSI/MSI-X
[  980.403409] nvidia 0000:b2:00.0: irq 286 for MSI/MSI-X
```

* This means that when _p2pBandwidthLatencyTest_ hangs, it's hanging because it's not getting a desired interrupt and doesn't run a necessary interrupt handler (seems plausible).
* [This changelog](https://tracker.debian.org/media/packages/n/nvidia-graphics-drivers/changelog-367.44-3) inspired me to try to disable MSI via driver module parm.  Specifically, the section below gave me the idea:

```
   - Changed the default PCIe interrupt delivery method from virtual-wire
      to MSI. Note that if the NVIDIA Linux driver fails to initialize
      with an error indicating that it is not receiving interrupts, MSI
      can be disabled by setting the module parameter "NVreg_EnableMSI=0"
      when loading the NVIDIA kernel module.
```
* No dice when setting `NVreg_EnableMSI=0` (note that I believe this also requires you to regenerate initramfs).
* In a hail mary attempt at trying to figure this out, I tried a newer version of the Nvidia driver (430.50) and a fresh install of CUDA 10.0.130 with no results, as expected (since we've already established that this isn't specific to a particular driver or CUDA version).

Basically, my workaround at this point is to tell the folks on my end that they need to wait 10 minutes (presumably the GPU memory has everything transferred to it within this time span) before killing a PyTorch process.

**Comment by ezyang** [comment]
Thank you for the detailed analysis. @jpellman, I agree with your analysis that this is not a PyTorch issue. Have you filed an nvbug for this yet?

**Comment by jpellman** [comment]
I just did.  You can find it [here](https://developer.nvidia.com/nvidia_bug/2712582) (NVBug 2712582).

**Comment by Dubrzr** [comment]
Hi @jpellman , thanks a lot!

I don't have access to NVBug, do you have any news ?

Thanks,
Julien

**Comment by jj0mst** [comment]
I encountered this issue too when I use DataParallel for 2 Tesla V100.

Strange thing:
-on a machine with SXM2, Intel CPU and V100 GPUs (16GB version) my training runs smoothly
-on another machine with PCIE, AMD CPU and V100 GPUs (32GB version) the training hangs

I just found that on 10/03 NV released these new drivers [https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-418-8701/index.html](https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-418-8701/index.html).

The first fixed issue seems similar to ours, but not exactly the same.

I can't look into NVBug to see the status either and I am not able to update the drivers for now.

@jpellman can you please give us some news?

**Comment by jpellman** [comment]
I've received no updates from Nvidia

**Comment by gottbrath** [comment]
@jpellman -- Thanks for the update. I'll ping my contacts at NVIDIA and make sure they are aware of this.

**Comment by apoms** [comment]
Any updates or workarounds? Observing this same issue with a DGX-1 (Tesla V100s) running CUDA 10, driver 430.50, and pytorch 1.3.0.

**Comment by natezhenghy** [comment]
Same issue. Pytorch 1.3, Driver 418.87, CUDA 10.1.

**Comment by jpellman** [comment]
I think the best workaround will be to fix #10657 :)

**Comment by leimao** [comment]
I am using `nn.DataParallel`. The training always work using PyTorch 1.2.0. But it froze and could only be killed using `sudo kill -09` when I was using PyTorch 1.3.0. Need to find out the source of the problem and a solution, without changing any of the PyTorch source code.

**Comment by rathken** [comment]
Turn off ACS in the BIOS or using setpci, ACS will prevent p2p for working. I have tested this on Linux with different AMD Ryzen 9 3900X systems with 2x RTX 2080 and 2X Quadro RTX 6000  with driver 410.79  and  cuda V10.0.130

turn ACS off will let p2pBandwidthLatencyTest run as it should.....

**Comment by jpellman** [comment]
Interesting- I had come across the Supermicro post [here](https://www.supermicro.com/support/faqs/faq.cfm?faq=20732) describing that potential issue, but when I ran `lspci -vvv | grep ACSCtrl` I didn't get any output.  I assumed that this meant that ACS wasn't a factor, but perhaps it's named something else on the particular host I'm working with.  Apparently this was the issue for someone else using Torch as well, and they also didn't receive any output from `grep` (https://github.com/twitter/torch-ipc/issues/17).

**Comment by jpellman** [comment]
ACS wasn't available in my BIOS.  There was an option/optimization to "Disable Unused PLX" that was enabled, but disabling that had no effect.  Just in case Supermicro stealthily enabled ACS without providing a BIOS option, I ran the `setpci` command provided:

```
for plx in $(lspci | grep -i plx | awk '{print $1}'); do sudo setpci -s ${plx} f2a.w=0000; done
```

That had no discernible effect too.  The only interesting thing is that I was able to clobber multiple _p2pBandwidthLatencyTest_ processes that used different numbers in their I/O syscalls:

```
tail -n 1 strace_p2pBandwidthLatencyTest-*
==> strace_p2pBandwidthLatencyTest-1 <==
ioctl(4, _IOC(0, 0x00, 0x42, 0x00)
==> strace_p2pBandwidthLatencyTest-2 <==
ioctl(4, _IOC(0, 0x00, 0x42, 0x00)
==> strace_p2pBandwidthLatencyTest-3 <==
ioctl(4, _IOC(0, 0x00, 0x21, 0x00)
==> strace_p2pBandwidthLatencyTest-4 <==
ioctl(4, _IOC(0, 0x00, 0x19, 0x00)(
```

`0x42` (66 in base 10) and `0x21` (33 in base 10) are new, but probably unrelated.  Overall, `nvidia-uvm` is till glitching somewhere.

**Comment by jpellman** [comment]
Actually, I missed the bit in the Supermicro knowledgebase where `setpci` wasn't persistent, and I think that I did a reboot in between when I ran the for loop in my last post and when I did the testing.  

I'm unable to reproduce the bug if I run the `setpci` for loop and then immediately start testing afterwards.  Since there doesn't seem to be a way to toggle ACS on/off in my BIOS (i.e., ACS is implicitly on by default despite Supermicro's claims that this isn't he case after BIOS 1.0c 2015/2/25) I'll have to find some way for `setpci` to execute upon boot for all the PLI devices (probably a udev rule).  For future reference, the mobo I've been working with is the SuperMicro [X11DPG-OT-CPU](https://www.supermicro.com/manuals/motherboard/C620/MNL-2017.pdf).

**Comment by rathken** [comment]
From the SuperMicro X11DPG-OT-CPU. "Select Enable to program Access Control Services (ACS) to Chipset PCI-E Root Port Bridges. Select Disable to program Access Control Services to all PCI-E Root Port Bridges.The options are Enable and Disable."  Are you saying that for this board it doesn't work? Just curious, I have several X10DRG-OT+-CPU and it work fine there. BIOS Version: 3.1 Release Date: 07/13/2018

**Comment by Dubrzr** [comment]
If this can provide more info, it also freezes with this motherboard : ProLiant XL270d Gen10.
Thanks

**Comment by rathken** [comment]
Is your ProLiant XL270d Gen10 with the NVlink tray or the PCIe tray?

**Comment by jpellman** [comment]
The "Select Enable to program Access Control Services (ACS)" option only shows up if VT-d is enabled, which it isn't on my system.  It would seem logical that since the ACS toggle is nested under the VT-d option (and disappears when VT-d is disabled), the ACS toggle would inherit the VT-d toggle's disabled state.  This doesn't seem to be the case, however.  This is confirmed by a fresh boot with my configuration (VT-d disabled, ACS enabled when VT-d enabled) indicating that the ACS register values are populated:

```
(base) [jsp2205@ax08 ~]$ for plx in $(lspci | grep -i plx | awk '{print $1}'); do sudo setpci -s ${plx} f2a.w; done
[sudo] password for jsp2205:
0000
001d
001d
0000
001d
001d
0000
001d
001d
0000
001d
001d
```

If I go into the UEFI/BIOS, temporarily enable VT-d to make the ACS toggle visible, disable ACS, and then toggle VT-d off again, I get the expected results:

```
(base) [jsp2205@ax08 ~]$ for plx in $(lspci | grep -i plx | awk '{print $1}'); do sudo setpci -s ${plx} f2a.w; done
0000
0000
0000
0000
0000
0000
0000
0000
0000
0000
0000
0000
```

Overall, seems to be a mixture of user error and interface confusion :)

**Comment by artemZholus** [comment]
Same problem. I've tested on Ryzen 3700 with two gtx 1060 6gb. `DataParallel` freezes at `forward` if using two gpus while behaving normally with one. `strace` also showed me deadlock at ioctl
```
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x4a, 0xc0), 0x7ffe8b783960) = 0
ioctl(7, _IOC(0, 0, 0x21, 0), 0x7ffe8b783380) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2a, 0x20), 0x7ffe8b7836b0) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2a, 0x20), 0x7ffe8b7836e0) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2a, 0x20), 0x7ffe8b7836c0) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2a, 0x20), 0x7ffe8b783520) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x2b, 0x20), 0x7ffe8b7836b0) = 0
ioctl(7, _IOC(0, 0, 0x1d, 0), 0x7ffe8b783710) = 0
ioctl(7, _IOC(0, 0, 0x21, 0), 0x7ffe8b7829a0) = 0
ioctl(6, _IOC(_IOC_READ|_IOC_WRITE, 0x46, 0x4a, 0xc0), 0x7ffe8b783960) = 0
ioctl(7, _IOC(0, 0, 0x21, 0), 0x7ffe8b783380) = 0
ioctl(7, _IOC(0, 0, 0x21, 0)
```
I'm using `PyTorch==1.3.0`, driver version `418.87` (also tried with `435.21` but unsuccessfully), `cuda==10.1`, `cudnn==7.6.5`

**Comment by artemZholus** [comment]
I managed to solve this issue in my case by disabling IOMMU in bios.

**Comment by BelBES** [comment]
I have the same issues with (2x1080ti + AMD CPU).
To solve this issue you should:

> 1. sudo nano /etc/default/grub
> 2. GRUB_CMDLINE_LINUX_DEFAULT="amd_iommu=soft"
> 3. sudo update-grub
> 4 .reboot

**Comment by y-x-c** [comment]
> I managed to solve this issue in my case by disabling IOMMU in bios.

Works for me!

**Comment by ezyang** [comment]
I'm downgrading the priority of this issue as it looks like people have found a workaround, and it seems there is not much we can do about it in PyTorch side. I'll keep it open for visibility, though.

**Comment by fpoms** [comment]
Can confirm the workaround of disabling IOMMU also works for me.

**Comment by Dubrzr** [comment]
Disabling IOMMU can be a security issue, by disabling it you authorize GPUs to access memory addresses of all other GPUs. 

This bug is probably in Nvidia drivers.
@jpellman : do you have any news on your NVBug post?

Thanks.

**Comment by jpellman** [comment]
RMA attacks aren't (at present) as much of a concern in our environment (the machines have very limited access and aren't publicly accessible), although it's certainly a far from an ideal scenario.  I closed out the NVBug post a while ago.  I'd re-open an issue with them and frame this as a security issue.

**Comment by Dubrzr** [comment]
I've posted an issue, here is the answer from Nvidia :

> Thanks you for reporting. I'm afraid IOMMU is not supported on Linux.
> As per https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#iommu-on-linux,"On Linux only, CUDA and the display driver does not support IOMMU-enabled bare-metal PCIe peer to peer memory copy. However, CUDA and the display driver does support IOMMU via VM pass through. As a consequence, users on Linux, when running on a native bare metal system, should disable the IOMMU. The IOMMU should be enabled and the VFIO driver be used as a PCIe pass through for virtual machines.
> On Windows the above limitation does not exist."

**Comment by jpellman** [comment]
¯\\_(ツ)_/¯ 

Lock up your Linux gamer workstations in a closet I guess.  If you're concerned about a DMA attack from some sort of rogue peripheral, it's probably easier to block peripheral ports in some way- either via epoxy resin on the low-tech end of things or at the firmware/driver level (I believe this is fairly common practice in fintech and some healthcare institutions).  It seems that there are a [couple other steps you could take](https://security.stackexchange.com/questions/88629/preventing-dma-attacks), even you can't do anything with IOMMU.

**Comment by fushuyue** [comment]
> I managed to solve this issue in my case by disabling IOMMU in bios.

Works for me. Many thanks!

**Comment by leolb-aphp** [comment]
> ¯\_(ツ)_/¯
> 
> Lock up your Linux gamer workstations in a closet I guess. If you're concerned about a DMA attack from some sort of rogue peripheral, it's probably easier to block peripheral ports in some way- either via epoxy resin on the low-tech end of things or at the firmware/driver level (I believe this is fairly common practice in fintech and some healthcare institutions). It seems that there are a [couple other steps you could take](https://security.stackexchange.com/questions/88629/preventing-dma-attacks), even you can't do anything with IOMMU.

The concern is not only about plugging a rogue device. It's also about the device itself failing to respect the host system security policy. We have no information on the device internals and how it enforces security constraints to user-submitted GPU workloads.

If not enforced correctly, a user submitted GPU task could write to kernel memory at arbitrary address and escalate privileges.

**Comment by leolb-aphp** [comment]
> Disabling IOMMU can be a security issue, by disabling it you authorize GPUs to access memory addresses of all other GPUs.
> 
> This bug is probably in Nvidia drivers.
> @jpellman : do you have any news on your NVBug post?
> 
> Thanks.

Not all other GPUs, but rather, all available main system memory on the system. GPUs have their own separate memory subsystem.

**Comment by vikramg1** [comment]
on my HP Envy dual boot laptop, I had to disable virtulization in the BIOS settings. Seems to work now.

**Comment by jramapuram** [comment]
Is there a better solution here than disabling IOMMU? Does upgrading CUDA/NCCL help?

**Comment by nfrumkin** [comment]
Try `NCCL_P2P_DISABLE=11 in front of your training script. See: https://github.com/pytorch/pytorch/issues/1637

Example: `NCCL_P2P_DISABLE=1 python3 train.py`

**Comment by oytunturk** [comment]
Thanks, setting NCCL_P2P_DISABLE=1 solved a similar issue I was having with RTX A5000s.

**Comment by Borodin** [comment]
I faced the same problem, all the solutions described here are not suitable for me.

I have AMD FX(tm)-4100 Quad-Core Processor
RAM: 8 GB
2x NVIDIA GeForce GTX 1060 6GB
PyTorch 1.13.1
Ubuntu 22.04.1 LTS

Any help is much appreciated!
